sandbox/indeed-scraping/indeed-scraper.py

```diff
@@ -0,0 +1,209 @@
+    locations.update(LOCATIONS_OTHER_STATES)
+    #start_url = "https://www.indeed.com/jobs?q=title%3A%28{}%29&l=Texas&radius=50&limit=50&start=50"
+    for location in locations:
+        try:
+            scrape_search_location(search_term, location, locations[location])
+        except Exception:
+            continue  # or you could use 'continue'
+        
+    log("Completed scraping {} jobs for Search:{}".format(len(unique_job_ids), search_term))
+ 
+def scrape_search_location(search_term, location, radius):
+    # sleep a random number of seconds
+    sleep_non_bot()
+    log("Search:{} in Location:{} within Radius:{}".format(search_term, location, str(radius)))
+    # URl with the query parameter for location, search term and limit. We do not use "start={}" parameter here. This is the starting point.
+    search_start_url = "https://www.indeed.com/jobs?q=title%3A%28{}%29&l={}&radius={}&limit=50".format(search_term, location, str(radius))
+    #scrape page 0
+    soup = scrape_jobs_page(search_start_url, search_term, location, 0)
+ 
+    #find how many more pages to scrape
+    search_counts = None
+    try:
+        search_counts = soup.find(id = 'searchCount').string.encode('utf-8').split() #this returns the total number of results
+    except Exception:
+        pass
+
+    if not search_counts:
+        log("Could not get the number of search counts")
+        return
+    
+    num_search_results = int(search_counts[3].replace(',',''))
+    log("Nummber of search results:{}".format(str(num_search_results)))
+    if num_search_results < 50:
+        log("No additional pages for Search:{} in Location:{} within Radius:{}".format(search_term, location, str(radius)))
+        return
+    
+    num_pages = num_search_results//50 + 1
+    log("Number of pages:{}".format(str(num_pages)))
+    
+    for page_number in range(1, num_pages):
+        sleep_non_bot()
+        start_from = 50 * page_number
+        #NOTE: the start={} is required here when we paginate.
+        search_url = search_start_url + "&start={}".format(str(start_from))
+        try:
+            scrape_jobs_page(search_url, search_term, location, page_number)
+        except Exception:
+            continue
+    
+
+def scrape_jobs_page(search_url, search_term, location, page_number):
+    log("Scraping URL:{}".format(search_url))
+    
+    search_page = requests.get(search_url)
+    soup = BeautifulSoup(search_page.text, "html.parser")
+    job_ids = extract_job_ids_from_page(soup)
+    
+    log("Number of job ids in the url:{} is {}".format(search_url, str(len(job_ids))))
+    
+    out_file_name = search_term + "_" + location + "_Page{}_".format(str(page_number)) + str(int(time.time()))    
+    log("Scraping jobs in the URL:{} into file:{}".format(search_url, out_file_name))
+    # save file locally and also to S3. One file contains all the job information (around 50) in one page
+    scrape_jobs(job_ids, out_file_name) 
+    data = open(out_file_name, 'rb')
```

Comment:
standard practice in Python is to use `with open()` so Python automagically handles closing file connections for you. e.g.:

```
with open(out_file_name, 'rb') as data:
    s3.Bucket(BUCKET_NAME).put_object(Key=s3_key, Body=data)
```

Reply:
^ I am recommending this change

Reply:
sure, makes sense, will do